pgxn/neon/communicator.c

```diff
@@ -2499,16 +2548,29 @@ communicator_read_slru_segment(SlruKind kind, int64 segno, neon_request_lsns *re
 void
 communicator_reconfigure_timeout_if_needed(void)
 {
-	bool	needs_set = MyPState->ring_receive != MyPState->ring_unused &&
+	bool	needs_set = (MyPState->ring_receive != MyPState->ring_unused ||
+						 (MIN_BACKEND_PREFETCH_LSN != InvalidXLogRecPtr &&
+						  MIN_BACKEND_PREFETCH_LSN != GetXLogReplayRecPtr(NULL))) &&
```

Comment:
This condition is extremely volatile and may be `true` a moment after it was determined `false`.  I don't think this condition is a valuable addition - at best it'll never be false, at worst it'll accidentally fail to set a timeout when it would be needed moments later.

Reply:
> Yes, this condition is volatile. But I do not see some problem in it.

The issue here is that if the condition is false (i.e., `MIN_BACKEND_PREFETCH_LSN == GetXLogReplayRecPtr(NULL)`), then needs_set may be false, which means the timer will not be set. If the cluster's replay LSN now changes, there won't be a timer that fires to signal that we need to update the LSN.

So, if you want to keep a non-invalid LSN in the backend's MIN_BACKEND_PREFETCH_LSN, you'll need to _always_ keep the timer active while the LSN is not invalid, not just "if it's not currently at the current replay LSN"

Reply:
Agree.
So I will just remove this condition: `MIN_BACKEND_PREFETCH_LSN == GetXLogReplayRecPtr(NULL)`
Alternatively we can assign `MIN_BACKEND_PREFETCH_LSN = InvalidXLogRecPtr` if prefetch ring is empty and `last_replay_lsn == InvalidXLogRecPtr`. But in this case we need to perform this double-check with memory barrier on next get_request_lsns. I think it will be better to keep timer always active.
What do you think?

Reply:
Ok, changed in this way.